ta4j-core/src/test/java/org/ta4j/core/analysis/criteria/AverageProfitCriterionTest.java

```diff
@@ -37,7 +37,10 @@
     private MockTimeSeries series;
 
     public AverageProfitCriterionTest(Function<Number, Num> numFunction) {
-        super(numFunction);
+        super(
+            (params) -> new AverageProfitCriterion(),
+            numFunction
+        );
```

Comment:
This lambda that discards `params` doesn't feel right.  I was able to pass a method reference `AverageProfitCriterion::new` but only after defining an explicit constructor:
```
    public AverageProfitCriterion(Object... params) {
    }
```
But that seems even worse.  This is not a requested change, just an item for discussion.

Reply:
Yeah, the solution I saw/see is just getting rid of the getCriterion altogether, or making it an abstract, although the latter solution will be almost the same in the end, as only 1 Crit requires params, all the others don't:

```java
    protected AverageProfitCriterionTest(Function<Number, Num> numFunction) {
        super(numFunction);
    }
    
    @Override
    protected AnalysisCriterion getCriterion(Object... params) {
        return new AverageProfitCriterion();
    }
```

or just plain:

```java
    protected AverageProfitCriterionTest(Function<Number, Num> numFunction) {
        super(numFunction);
    }

    protected AnalysisCriterion getCriterion() {
        return new AverageProfitCriterion();
    }
```

Reply:
The factory was added so that the unit tests can call `getCriterion(Object... params)` consistently in all `Criterion` unit tests.  Also, when implementing a new `Criterion's` unit test you can select an appropriate "parent" unit test, copy it, rename it, and make it work with minimal editing.  Otherwise you'd have to refactor the name of the `Criterion`.  It's also easier to `diff` two different `Criteria` unit tests.  Finally we have the ability to develop a generic unit test framework that can work on all `Criteria`.

See [XLS Testing](https://ta4j.github.io/ta4j-wiki/XLS-Testing.html) and read the `Copy-Paste-Edit Quickstart` section.  Looks I may have to edit some of the formatting on that page..

Reply:
Ah I see, in practice though it doesn't really work out though, I diffed some of the test classes, but the tests are pretty different, the only thing that remains the same is `betterThan`, the rest is pretty different.

The lambda that disregards is the parameters is the result of having the 'abstract' getCriterion. I really had to look and see how it's working the first time I touched on of these tests. I think it's easier to understand if all the test classes are just like this:

```    
public class AverageProfitCriterionTest extends AbstractCriterionTest{
    private Analysis criterion;

    protected AverageProfitCriterionTest(Function<Number, Num> numFunction) {
        super(numFunction);
    }

    @Before
    protected void createCriterion() {
        criterion = new AverageProfitCriterion();
    }

    // ...
}
```

If all of them are like this, they will c/p exactly this from another one, and we'll get the same result in the end, consistent tests.

But i'm fine to let it rest, and let it be/merge it like this as well.

(Side note, still working on sqrt, mainly with making a good implementation for the initial guess, and trying to understand that algebra in the wiki :))

Reply:
Actually, I have another approach, see the last commit. It's up for discussion.

Reply:
Okay, I'll take a look today.  If there's a better way, I'm all for it.